text/unification.md

```diff
@@ -0,0 +1,268 @@
+
+## Simple unification problems
+
+We call *simple unification problem* an equation of the form `Γ ⊢ ?e@{subst} =?= t` or an inequation of the form `Γ ⊢ ?e@{subst} ≤? t`. Unification basically reduces arbitrary complex unification problems to simple ones which in turn may generate new arbitrary complex problems.
+
+## Simplification of reflexive unification problems
+
+A reflexive unification problem induces constraints on the possible definitions of `?e`: once `?e` is defined, it may generate new unification problems between terms of the instances that contain existential variables.
+
+A reflexive unification problem can be simplified as follows:
+- Variables of the context whose instances in the substitution are provably incompatible (e.g. because corresponding to distinct head normal forms) can be removed from the context.
+- If both instances are free of existential variables, the problem can be dropped since its resolution will eventually only create equations between convertible existential-variables-free terms.
+
+## Simplification of atomic unification problems
+
+An atomic unification problem can be simplified as follows:
+- Variables of the context of the existential variables that are bound to terms mentioning variables of the context of the equations in non-erasable position in only one of the two instances can be restricted since there would be no way to make these terms match a subterm of the instance of the other existential variables; otherwise said, both existential variables can be restricted on the (well-typed closure of the) subsets of their context which intersect.
+- If the problem is an inequation, it can only be postponed since one variable could be instantiated e.g. with `Prop` and the other with `Type` without them to be equal.
+- If one of the instances mention only distinct variables on the subset where there is an intersection, *and* the problem is not an inequation, the corresponding existential variable can be defined with the other one without loss of solutions
+
+## Basic algorithm for resolving simple unification problems
+
+Solving non atomic simple unification problems relies on basically two concepts (see Huet 1975, `MATCH` algorithm):
+- *Projection*, that is, solving `Γ ⊢ ?e@{a:=t,b:=t} =?= t` by either setting `Δ ⊢ ?e := a` or `Δ ⊢ ?e := b` so that `?e@{a:=t,b:=t}` indeed evaluates in `Γ` to `t`.
+
+  In the more general case, a projection requires to solve general unification problems, that is, to know if `Γ ⊢ ?e@{a:=t} =?= u` can be solved by projecting `a` requires to find a context `E` such that `Γ ⊢ E[t] =?= u` is resolvable (up to reduction of `E[t]`). This can be done in a systematic way in algebraic-type-free λ-calculus by reasoning on the normal forms of `t` and `u` and using type constraints. It is not clear however that a systematic study of the process and of the form of evaluation ontexts has already been done somewhere in the presence of algebraic types.
```

Comment:
```suggestion
  In the more general case, a projection requires to solve general unification problems, that is, to know if `Γ ⊢ ?e@{a:=t} =?= u` can be solved by projecting `a` requires to find a context `E` such that `Γ ⊢ E[t] =?= u` is resolvable (up to reduction of `E[t]`). This can be done in a systematic way in algebraic-type-free λ-calculus by reasoning on the normal forms of `t` and `u` and using type constraints. It is not clear however that a systematic study of the process and of the form of evaluation contexts has already been done somewhere in the presence of algebraic types.
```